### PR TITLE
Alerting: Introduce Metric Aggregation starting with Silences

### PR DESCRIPTION
**Pull request by gotjosh**
This PR introduces the exposition of the individual Alertmanager metrics, it starts with Silences as a baseline so that the reviewer can get a better feeling of what's going on without overwhelm but expect the next set right after.

Introduces the following metrics:

```
grafana_alerting_silences_gc_duration_seconds
grafana_alerting_silences_snapshot_duration_seconds
grafana_alerting_silences_snapshot_size_bytes
grafana_alerting_silences_queries_total
grafana_alerting_silences_query_errors_total
grafana_alerting_silences_query_duration_seconds
grafana_alerting_silences_gossip_messages_propagated_total
grafana_alerting_silences
```

### CODE REVIEW

File: pkg/services/ngalert/metrics/multi_org_alertmanager.go [hunk 1]
```diff
@@ -28,5 +37,112 @@ func NewMultiOrgAlertmanagerMetrics(r prometheus.Registerer) *MultiOrgAlertmanag
...
+		return reg
+	}
+
+	result := prometheus.NewRegistry()
+	moa.registries.AddUserRegistry(sid, result)
+
+	return result
+}
+
+// AlertmanagerAggregatedMetrics are metrics collected directly from the registry.
```
Comment (gotjosh) [inline pkg/services/ngalert/metrics/multi_org_alertmanager.go:69]: I've created a separate struct as I have close to 30 other metrics that I will include next.

It felt a bit confusing to have `metrics.Alertmanager` and another struct that that reads "some form of alertmanager metrics" but I ended up settling for `AlertmanagerAggregatedMetrics` as this is exactly what it is doing.

File: pkg/services/ngalert/metrics/multi_org_alertmanager.go [hunk 2]
```diff
@@ -28,5 +37,112 @@ func NewMultiOrgAlertmanagerMetrics(r prometheus.Registerer) *MultiOrgAlertmanag
...
+	out <- a.silencesQueryErrorsTotal
+	out <- a.silencesQueryDuration
+	out <- a.silencesPropagatedMessagesTotal
+	out <- a.silences
+}
+
+func (a *AlertmanagerAggregatedMetrics) Collect(out chan<- prometheus.Metric) {
+	data := a.registries.BuildMetricFamiliesPerUser()
+
+	data.SendSumOfSummaries(out, a.silencesGCDuration, "alertmanager_silences_gc_duration_seconds")
```
Comment (gotjosh) [inline pkg/services/ngalert/metrics/multi_org_alertmanager.go:140]: You can go read the function signature to get a gist of what's going but it gets hairy please quickly.

Of each `SendSum` function there's two kinds, one with a suffix of `PerUser` and one without the suffix. These allow us to control if you want to perform an aggregation (and effectively turn a metric that will be per user to per instance).

For example, say we have `alertmanager_silences_snapshot_size_bytes` and 3 Grafana tenants. With the per user metric we would get:

```
alertmanager_silences_snapshot_size_bytes{org="1"} 10
alertmanager_silences_snapshot_size_bytes{org="2"} 20
alertmanager_silences_snapshot_size_bytes{org="3"} 30
```
With the function that does not have the suffix we get:

```
alertmanager_silences_snapshot_size_bytes 50
```

This technique helps us reduce per-tenant cardinality. We did a pretty extensive analysis in Mimir back when we implemented this so I just copied the work that we did - that being said, there are probably changes that we'll make over time as we discover what kind of balance works for Grafana.

File: pkg/services/ngalert/metrics/multi_org_alertmanager.go [hunk 3]
```diff
@@ -1,21 +1,30 @@
...
 type MultiOrgAlertmanager struct {
-	Registerer               prometheus.Registerer
+	Registerer prometheus.Registerer
+	registries *UserRegistries
+
 	ActiveConfigurations     prometheus.Gauge
 	DiscoveredConfigurations prometheus.Gauge
-	registries               *OrgRegistries
+
+	aggregatedMetrics *AlertmanagerAggregatedMetrics
```
Comment (gotjosh) [inline pkg/services/ngalert/metrics/multi_org_alertmanager.go:20]: You're probably wondering where is the test for this - well, it's complicated.

To do the testing, I'd have to basically "copy" all of the metrics from the Alertmanager just for the test setup - it feels like we're entering the land of diminishing returns and have opted out from doing anything like that right now.

If you feel strongly about having them I can probably figure something out at the expense of more time.

**Review by alexweav — APPROVED** [review]

File: pkg/services/ngalert/metrics/multi_org_alertmanager.go [hunk 4]
```diff
@@ -28,5 +37,112 @@ func NewMultiOrgAlertmanagerMetrics(r prometheus.Registerer) *MultiOrgAlertmanag
...
+			fmt.Sprintf("%s_%s_silences_snapshot_size_bytes", Namespace, Subsystem),
+			"Size of the last silence snapshot in bytes.",
+			nil, nil),
+		silencesQueriesTotal: prometheus.NewDesc(
+			fmt.Sprintf("%s_%s_silences_queries_total", Namespace, Subsystem),
+			"How many silence queries were received.",
+			nil, nil),
+		silencesQueryErrorsTotal: prometheus.NewDesc(
+			fmt.Sprintf("%s_%s_silences_query_errors_total", Namespace, Subsystem),
+			"How many silence received queries did not succeed.",
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/metrics/multi_org_alertmanager.go:107]: You wanna open a PR for that upstream?

File: pkg/services/ngalert/metrics/multi_org_alertmanager.go [hunk 5]
```diff
@@ -28,5 +37,112 @@ func NewMultiOrgAlertmanagerMetrics(r prometheus.Registerer) *MultiOrgAlertmanag
...
+
+	return moa
+}
+
+// RemoveOrgRegistry removes the *prometheus.Registry for the specified org. It is safe to call concurrently.
+func (moa *MultiOrgAlertmanager) RemoveOrgRegistry(id int64) {
+	moa.registries.RemoveUserRegistry(strconv.FormatInt(id, 10), false)
+}
+
+// GetOrCreateOrgRegistry gets or creates a *prometheus.Registry for the specified org. It is safe to call concurrently.
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/metrics/multi_org_alertmanager.go:55]: We're going for a "soft deletion", that's the boolean on the remove registry call, so metrics are never removed just the reference to the user.

File: pkg/services/ngalert/metrics/multi_org_alertmanager.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/metrics/multi_org_alertmanager.go:107]: I can merge and probably commit this suggestion - two in one go.

**Review by alexweav — APPROVED** [review]

**Pull request merged**